modAionImpl/src/org/aion/zero/impl/db/RecoveryUtils.java

```diff
@@ -364,4 +376,142 @@ public static void pruneOrRecoverState(String pruning_type) {
+     * not perform the checks for side chains.
+     *
+     * <p>The minimum start height is 0, i.e. the genesis block. Specifying a height can be useful
+     * in performing the operation is sessions.
+     *
+     * @param startHeight the height from which to start importing the blocks
+     * @implNote The assumption is that the stored blocks are correct, but the code may interpret
+     *     them differently.
+     */
+    public static void reimportMainChain(long startHeight) {
+        // ensure mining is disabled
+        CfgAion cfg = CfgAion.inst();
+        cfg.dbFromXML();
+        cfg.getConsensus().setMining(false);
+        cfg.getDb().setHeapCacheEnabled(false);
+
+        System.out.println("\nRe-importing stored blocks INITIATED...\n");
+
+        Map<String, String> cfgLog = new HashMap<>();
+        cfgLog.put("GEN", "INFO");
+        cfgLog.put("DB", "WARN");
+        cfgLog.put("CONS", "ERROR");
+
+        AionLoggerFactory.init(cfgLog);
+
+        AionBlockchainImpl chain = AionBlockchainImpl.inst();
+        AionRepositoryImpl repo = chain.getRepository();
+        AionBlockStore store = repo.getBlockStore();
+
+        // determine the parameters of the rebuild
+        AionBlock block = store.getBestBlock();
+        AionBlock startBlock;
+        long currentBlock;
+        if (startHeight <= block.getNumber()) {
+            System.out.println(
+                    "\nRe-importing the main chain from block #"
+                            + startHeight
+                            + " to block #"
+                            + block.getNumber()
+                            + ". This may take a while.\n"
+                            + "The time estimates are optimistic based on current progress.\n"
+                            + "It is expected that later blocks take a longer time to import due to the increasing size of the database.\n");
+
+            if (startHeight == 0L) {
+                // dropping databases that can be inferred when starting from genesis
+                List<String> keep = List.of("block", "index");
+                repo.dropDatabasesExcept(keep);
+
+                // recover genesis
+                AionGenesis genesis = cfg.getGenesis();
+                AionHubUtils.buildGenesis(genesis, repo);
+                System.out.println("\nFinished rebuilding genesis block.");
+                startBlock = genesis;
+                currentBlock = 1L;
+            } else {
+                startBlock = store.getChainBlockByNumber(startHeight - 1);
+                currentBlock = startHeight;
+            }
+
+            chain.setBestBlock(startBlock);
+
+            long topBlockNumber = block.getNumber();
+            long stepSize = 10_000L;
+
+            Pair<ImportResult, AionBlockSummary> result;
+            final int THOUSAND_MS = 1000;
+
+            boolean fail = false;
+
+            long start = System.currentTimeMillis();
```

Comment:
nanoTime is more recommended for calculating time differences, I believe 

https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()

Reply:
Since the estimates are very inaccurate due to the nature of the operations being timed, I don't think switching to `nanoTime` would be very useful here. Especially since at best they give a minimum wait time to completion estimate, which is typically overly optimistic.

Reply:
I'll make a note of this suggestion in the initial issue for later enhancement.